diff -ru ucarp-1.5-pristine/src/carp.c ucarp-1.5/src/carp.c
--- ucarp-1.5-pristine/src/carp.c	2008-07-31 00:57:44.000000000 -0700
+++ ucarp-1.5/src/carp.c	2008-08-07 16:29:01.000000000 -0700
@@ -242,12 +244,21 @@
     eh.ether_shost[4] = 0x00;
     eh.ether_shost[5] = vhid;
     
-    eh.ether_dhost[0] = 0x01;
-    eh.ether_dhost[1] = 0x00;
-    eh.ether_dhost[2] = 0x5e;
-    eh.ether_dhost[3] = 0x00;
-    eh.ether_dhost[4] = 0x00;
-    eh.ether_dhost[5] = 0x12;        
+    if (no_mcast) {
+        eh.ether_dhost[0] = 0xff;
+        eh.ether_dhost[1] = 0xff;
+        eh.ether_dhost[2] = 0xff;
+        eh.ether_dhost[3] = 0xff;
+        eh.ether_dhost[4] = 0xff;
+        eh.ether_dhost[5] = 0xff;        
+    } else {
+        eh.ether_dhost[0] = 0x01;
+        eh.ether_dhost[1] = 0x00;
+        eh.ether_dhost[2] = 0x5e;
+        eh.ether_dhost[3] = 0x00;
+        eh.ether_dhost[4] = 0x00;
+        eh.ether_dhost[5] = 0x12;        
+    }
     
     eh.ether_type = htons(ETHERTYPE_IP);    
     
@@ -757,15 +773,17 @@
                 interface == NULL ? "-" : interface, strerror(errno));
         return -1;
     }
-    memset(&req_add, 0, sizeof req_add);
-    req_add.imr_multiaddr.s_addr = inet_addr("224.0.0.18");
-    req_add.imr_interface.s_addr = srcip.s_addr;
-    if (setsockopt(fd, IPPROTO_IP, IP_ADD_MEMBERSHIP,
-		   &req_add, sizeof req_add) < 0) {
-	logfile(LOG_ERR, "Can't do IP_ADD_MEMBERSHIP errno=%s (%d)",
-		strerror(errno), errno);
-	close(fd);
-	return -1;
+    if (!no_mcast) {
+        memset(&req_add, 0, sizeof req_add);
+        req_add.imr_multiaddr.s_addr = inet_addr("224.0.0.18");
+        req_add.imr_interface.s_addr = srcip.s_addr;
+        if (setsockopt(fd, IPPROTO_IP, IP_ADD_MEMBERSHIP,
+		       &req_add, sizeof req_add) < 0) {
+	    logfile(LOG_ERR, "Can't do IP_ADD_MEMBERSHIP errno=%s (%d)",
+		    strerror(errno), errno);
+	    close(fd);
+	    return -1;
+        }
     }
 #ifdef SIOCGIFFLAGS    
     if (strlen(interface) >= sizeof iface.ifr_name) {
diff -ru ucarp-1.5-pristine/src/globals.h ucarp-1.5/src/globals.h
--- ucarp-1.5-pristine/src/globals.h	2008-05-06 03:44:14.000000000 -0700
+++ ucarp-1.5/src/globals.h	2008-08-07 16:27:01.000000000 -0700
@@ -26,6 +26,7 @@
 GLOBAL0(signed char no_syslog);
 GLOBAL0(signed char daemonize);
 GLOBAL0(signed char ignoreifstate);
+GLOBAL0(signed char no_mcast);
 GLOBAL(int syslog_facility, DEFAULT_FACILITY);
 GLOBAL0(char *vaddr_arg);
 GLOBAL0(char *xparam);
diff -ru ucarp-1.5-pristine/src/ucarp.c ucarp-1.5/src/ucarp.c
--- ucarp-1.5-pristine/src/ucarp.c	2008-05-06 03:46:57.000000000 -0700
+++ ucarp-1.5/src/ucarp.c	2008-08-07 16:27:01.000000000 -0700
@@ -40,6 +40,7 @@
 	    "--shutdown (-z): call shutdown script at exit\n"
 	    "--daemonize (-B): run in background\n"
 	    "--ignoreifstate (-S): ignore interface state (down, no carrier)\n"
+	    "--nomcast (-M): use broadcast (instead of multicast) advertisements\n"
 	    "--facility=<facility> (-f): set syslog facility (default=daemon)\n"
 	    "--xparam=<value> (-x): extra parameter to send to up/down scripts\n"	    
 	    "\n"
@@ -233,6 +234,10 @@
             }
             break;
         }
+	case 'M': {
+	    no_mcast = 1;
+	    break;
+	}
         default: {
             usage();
         }
diff -ru ucarp-1.5-pristine/src/ucarp_p.h ucarp-1.5/src/ucarp_p.h
--- ucarp-1.5-pristine/src/ucarp_p.h	2008-05-06 03:48:39.000000000 -0700
+++ ucarp-1.5/src/ucarp_p.h	2008-08-07 16:27:01.000000000 -0700
@@ -1,7 +1,7 @@
 #ifndef __CARP_P_H__
 #define __CARP_P_H__ 1
 
-static const char *GETOPT_OPTIONS = "i:s:v:p:Pa:hb:k:x:nu:d:r:zf:BoS";
+static const char *GETOPT_OPTIONS = "i:s:v:p:Pa:hb:k:x:nu:d:r:zf:BoSM";
 
 static struct option long_options[] = {
     { "interface", 1, NULL, 'i' },
@@ -21,6 +21,7 @@
     { "facility", 1, NULL, 'f' },
     { "daemonize", 0, NULL, 'B' },
     { "ignoreifstate", 0, NULL, 'S' },
+    { "nomcast", 0, NULL, 'M' },
     { "passfile", 1, NULL, 'o' },
     { "xparam", 1, NULL, 'x' },
     { NULL, 0, NULL, 0 }
