diff -ru ucarp-1.4.pristine/src/carp.c ucarp-1.4/src/carp.c
--- ucarp-1.4.pristine/src/carp.c	2008-01-19 14:27:13.000000000 -0800
+++ ucarp-1.4/src/carp.c	2008-01-29 17:25:13.000000000 -0800
@@ -242,13 +242,13 @@
     eh.ether_shost[4] = 0x00;
     eh.ether_shost[5] = vhid;
     
-    eh.ether_dhost[0] = 0x01;
-    eh.ether_dhost[1] = 0x00;
-    eh.ether_dhost[2] = 0x5e;
-    eh.ether_dhost[3] = 0x00;
-    eh.ether_dhost[4] = 0x00;
-    eh.ether_dhost[5] = 0x12;        
-    
+    eh.ether_dhost[0] = 0xff;
+    eh.ether_dhost[1] = 0xff;
+    eh.ether_dhost[2] = 0xff;
+    eh.ether_dhost[3] = 0xff;
+    eh.ether_dhost[4] = 0xff;
+    eh.ether_dhost[5] = 0xff;        
+
     eh.ether_type = htons(ETHERTYPE_IP);    
     
     memcpy(pkt, &eh, sizeof eh);
