--- ucarp-1.3.pristine/src/carp.c	2007-07-04 08:17:12.000000000 -0700
+++ ucarp-1.3/src/carp.c	2007-12-18 16:10:43.000000000 -0800
@@ -647,6 +647,7 @@
     struct ifreq iface;
     int fd;
     int nfds;
+    int iface_running = 1;
 
     sc.sc_vhid = vhid;
     sc.sc_advbase = advbase;
@@ -735,9 +736,17 @@
 	    sc.sc_ad_tmo.tv_usec = 0;
 	    sc.sc_md_tmo.tv_sec = 0;
 	    sc.sc_md6_tmo.tv_usec = 0;
+            if (iface_running) {
+                iface_running = 0;
+            }
 	    sleep(SECONDS_TO_WAIT_AFTER_INTERFACE_IS_DOWN);
 	    continue;
-	}
+	} else {
+            if (!iface_running) {
+                iface_running = 1;
+                carp_setrun(&sc, 0);
+            }
+        }
 #endif
         nfds = poll(pfds, (nfds_t) 1, sc.sc_advbase * 1000);
         if (nfds == -1) {
